Initial commit

AzGoalie opened the pull request:
#### What does this PR do?
Initial commit of openddms.
#### Who is reviewing it? 
@coyotesqrl @bakejeyner @ahoffer @brendan-hofmann @andrewkfiedler @gordocanchola @brjeter

Commit:
Initial commit

mackncheesiest commented on src/main/kotlin/org/codice/ddms/gml/v3/SrsAttributes.kt:
```diff
@@ -0,0 +1,41 @@
+    val uomLabels: List<String> = emptyList()
+) {
+    override fun toString(): String {
+        val sb = StringBuffer()
```
❓ Any reason for `StringBuffer` here and `StringBuilder` [here](https://github.com/codice/openddms/pull/1/files#diff-d84138f8416732a12307b29fef846e2dR21)?

AzGoalie replied:
Nope, good catch 👍

mackncheesiest reviewed (commented):
Asking the hard questions 😄

Commit:
Changed StringBuffer to StringBuilder

coyotesqrl commented on build.gradle.kts:
```diff
@@ -0,0 +1,49 @@
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+plugins {
+    kotlin("jvm") version "1.2.60"
```
✏️ Should bump to latest, `1.2.61` as of this review.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/DateFormat.kt:
```diff
@@ -0,0 +1,77 @@
+    const val NOT_APPLICABLE = "Not Applicable"
+    private val formatters = listOf(
+            DateTimeFormatter.ISO_OFFSET_DATE_TIME,             // yyyy-MM-dd'T'hh:mm:ss.sTZD
+            DateTimeFormatter.ISO_LOCAL_DATE,                   // yyy-MM-dd
```
✏️ The comment should be `yyyy-MM-dd`.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/DateFormat.kt:
```diff
@@ -0,0 +1,77 @@
+        if (unknownOrNotApplicable && (date == UNKNOWN || date == NOT_APPLICABLE))
+            return true
+
+        for (formatter in formatters) {
```
✏️ This is potentially very inefficient/expensive (depending on how common any of the date formats other that ISO_OFFSET are. Since the four formats this accepts are of distinct sizes, it might be better to check the length of the input `date` and select the one formatter that _might_ be able to interpret the value. If that one can't, then you can catch/ignore and return false.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/DateFormat.kt:
```diff
@@ -0,0 +1,77 @@
+        return false
+    }
+
+    fun parse(date: String): Date? {
```
✏️ Similar inefficiency here. If you extract some of the logic from `isValid` (excluding the allowance for `unknownOrNotApplicable` values) you can create a simple `getParser` private method that returns a small data object containing the parser that knows how to convert to date. In the above method, a return value of `null` would mean you can't parse and you can return `false`. In this method, you can just return the Date off of that decorated instance.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/Position.kt:
```diff
@@ -0,0 +1,30 @@
+    val srsAttributes: SrsAttributes = SrsAttributes()
+) {
+    override fun toString(): String {
+        val sb = StringBuilder()
```
✏️ More Kotlin-y (and a bit easier to read):
```kotlin
        return """<gml:pos $srsAttributes xmlns:gml="http://www.opengis.net/gml">
            |${points.joinToString(" ")}
            |"</gml:pos>
        """.trimMargin().replace("\n", "")
```

(Double check that the output is what you want it to be, of course. The `\n` replacement is annoying, but putting it all on one line would be too long.)

AzGoalie replied:
Done, but it's still a little funky because we need to account for when `srsAttributes` are empty else there would be a double space.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/SrsAttributes.kt:
```diff
@@ -0,0 +1,41 @@
+    val axisLabels: List<String> = emptyList(),
+    val uomLabels: List<String> = emptyList()
+) {
+    override fun toString(): String {
```
✏️ Note that with the current implementation, you're inserting a blank space before everything but `srsName`, even if there is no `srsName`. A comparable, slightly more functional version of this would be...
```kotlin
    private fun srsNameToString(): String {
        return if (srsName.isNotBlank()) """srsName="$srsName"""
        else ""
    }

    private fun srsDimensionToString(): String {
        return if (srsDimension > 0) """srsDimension=$srsDimension"""
        else ""
    }

    private fun axisLabelsToString(): String {
        return if (axisLabels.any(String::isNotBlank))
            """axisLabels="${axisLabels.joinToString(" ")}""""
        else ""
    }

    private fun uomLabelsToString(): String {
        return if (uomLabels.any(String::isNotBlank))
            """uomLabels="${uomLabels.joinToString(" ")}""""
        else ""
    }

    override fun toString(): String {
        return """
            |${srsNameToString()}
            | ${srsDimensionToString()}
            | ${axisLabelsToString()}
            | ${uomLabelsToString()}
        """.trimMargin().replace("\n", " ")
    }
```

I don't think, in this case, the raw string + functions is better, but it is a bit more testable.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/builder/LinearRingBuilder.kt:
```diff
@@ -0,0 +1,42 @@
+import org.codice.ddms.gml.v3.LinearRing
+import org.codice.ddms.gml.v3.Position
+
+fun linearRing(init: LinearRingBuilder.() -> Unit) =
```
✏️ As a personal preference, I don't like to pollute the global namespace too much. Would suggest moving `linearRing` (and similarly with all the other functions supporting the DSL) into companion object of class. Then consumers would just invoke `LinearRingBuilder.linearRing()` or an import could be added for `import org.codice.ddms.gml.v3.builder.LinearRingBuilder.Companion.linearRing`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/builder/PolygonBuilder.kt:
```diff
@@ -0,0 +1,54 @@
+fun polygon(init: PolygonBuilder.() -> Unit) = PolygonBuilder().apply(init).build()
+
+class PolygonBuilder : Builder<Polygon> {
+    private lateinit var id: String
```
❓ I see that in `PointBuilder`, `id` is being initialized to an empty string instead of being lateinit. Is there a reason they're treated differently?

AzGoalie replied:
Nope, good catch

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/builder/PositionBuilder.kt:
```diff
@@ -0,0 +1,47 @@
+    private val points: ArrayList<Double> = arrayListOf()
+    private var srsAttributes: SrsAttributes = SrsAttributes()
+
+    fun points(vararg point: Double): PositionBuilder {
```
❓ Don't these doubles need to be a pair to represent a point? If so, should enforce that through the signature or at least with a `require` check.

AzGoalie replied:
It changes based on the srsName, i don't think we should have an enforcement here even though the most common are pairs and trios of doubles.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/builder/PositionBuilder.kt:
```diff
@@ -0,0 +1,47 @@
+        return this
+    }
+
+    fun points(point: List<Double>): PositionBuilder {
```
❓ Same here: don't these doubles need to come in pairs?

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/builder/SrsAttributesBuilder.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    fun srsDimension(srsDimension: Int): SrsAttributesBuilder {
+        this.srsDimension = srsDimension
```
❓ Are there any restrictions on dimensions? Min/max size?

AzGoalie replied:
Just that it needs to be positive, added a require check to `SrsAttributes`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/builder/SrsAttributesBuilder.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    fun axisLabels(vararg axisLabel: String): SrsAttributesBuilder {
+        axisLabels.addAll(axisLabel)
```
❓ Are there any restrictions on the labels? Blank/null values?

AzGoalie replied:
We filter out empty values when we print. Could filter them out here when we add them but that doesn't stop someone from manually creating a `SrsAttributes` with empty strings.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/Ddms20Resource.kt:
```diff
@@ -0,0 +1,68 @@
+    override val relatedResources: List<RelatedResources> = emptyList()
+) : DdmsResource {
+    init {
+        require(identifiers.isNotEmpty()) {
```
✏️ Should this instead confirm that there is at least one, non-blank identifier?

AzGoalie replied:
Added validation to check for non blank values in `Identifier`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/Ddms20Resource.kt:
```diff
@@ -0,0 +1,68 @@
+        require(identifiers.isNotEmpty()) {
+            "ddms:Resource must contain at least one ddms:identifier"
+        }
+        require(titles.isNotEmpty()) {
```
✏️ Should this instead confirm that there is at least one, non-blank title?

AzGoalie replied:
Added extra validation to check for non blank value in `Title`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/Ddms20Resource.kt:
```diff
@@ -0,0 +1,68 @@
+        require(titles.isNotEmpty()) {
+            "ddms:Resource must contain at least one ddms:title"
+        }
+        require(creators.isNotEmpty() || publishers.isNotEmpty() ||
```
✏️ Similar question here, where instead of only checking that at least one of the lists has an entry, we confirm that at least one of the lists contains a non-blank entry?

AzGoalie replied:
`Creator` `Publisher` `Contributor` and `PointOfContact` all have a requirement that there isn't a blank name, should cover this case.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/ContactBuilder.kt:
```diff
@@ -0,0 +1,61 @@
+        return this
+    }
+
+    fun person(init: PersonBuilder.() -> Unit): ContactBuilder {
```
❓ Function should be an overload of the `producer` name, yes?

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/ContactBuilder.kt:
```diff
@@ -0,0 +1,61 @@
+        return this
+    }
+
+    fun organization(init: OrganizationBuilder.() -> Unit): ContactBuilder {
```
I see the structure now, where `producer` is being set to a producer, person, organization, or service as appropriate. Not sure whether it's better to have the functions all named as they are now or to overload the `producer` function with the different value/builder inputs. I don't feel too strongly either way.

AzGoalie replied:
I'm open to either option, going to leave it as is for now unless there's a compelling reason to change

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/Ddms20ResourceBuilder.kt:
```diff
@@ -0,0 +1,350 @@
+        Ddms20ResourceBuilder().apply(init).build()
+
+class Ddms20ResourceBuilder : Builder<Ddms20Resource> {
+    private val identifiers: ArrayList<Identifier> = arrayListOf()
```
✏️ I haven't scrolled to confirm, but I suspect these fields are all ordered the same as in the `Ddms20Resource` in which case I understand the ordering but still don't love it. I think it would be easier for another engineer coming to this later to see all the fields ordered/grouped logically:
* all vals
* all vars
* all lateinit vars

AzGoalie replied:
I had them ordered to make it easier to confirm that i wasn't missing anything from the schemas. Now that there are proper tests, i can rearrange them.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/GeospatialCoverageBuilder.kt:
```diff
@@ -0,0 +1,91 @@
+    private val postalAddresses: ArrayList<PostalAddress> = arrayListOf()
+    private val verticalExtents: ArrayList<VerticalExtent> = arrayListOf()
+
+    fun geographicIdentifiers(vararg geographicIdentifier: GeographicIdentifier): GeospatialCoverageBuilder {
```
❓ Do you not want/need to support an overload that takes a `List<GeographicIdentifier>`? The other builders all have the `List` overloads when they have `vararg`.

AzGoalie replied:
Must have been an oversight, good catch

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/GeospatialCoverageBuilder.kt:
```diff
@@ -0,0 +1,91 @@
+        return geographicIdentifiers(geographicIdentifier)
+    }
+
+    fun boundingBoxes(vararg boundingBox: BoundingBox): GeospatialCoverageBuilder {
```
❓ Same question about overloading with `List` flavor of fun.

AzGoalie replied:
Must have been an oversight, good catch

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/GeospatialCoverageBuilder.kt:
```diff
@@ -0,0 +1,91 @@
+        return this
+    }
+
+    fun boundingBox(west: Double, east: Double, north: Double, south: Double) =
```
❓ Is this a usual/canonical ordering for the four compass points in ddms? Because it reads weird to me. I expect to see NSEW-ordered lists when there is no external ordering requirement.

❓ How do these four doubles define a box? Each point on a plane (or virtual plane) should require two coordinates to define it, not one. This seems to relate to my earlier question about points. Is there some translation in ddms that can determine a point from a single double?

AzGoalie replied:
I can change the ordering, i forgot why i had it this way.
As for how this defines a box, it creates a box with with North / South being the northern and southern most Latitude values and East / West  being the eastern and western most Longitude values.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/GeospatialCoverageBuilder.kt:
```diff
@@ -0,0 +1,91 @@
+        return boundingGeometries(boundingGeometry)
+    }
+
+    fun postalAddresses(vararg postalAddress: PostalAddress): GeospatialCoverageBuilder {
```
❓ Same question about List overload.

AzGoalie replied:
Must have been an oversight, good catch

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/GeospatialCoverageBuilder.kt:
```diff
@@ -0,0 +1,91 @@
+        return postalAddresses(postalAddress)
+    }
+
+    fun verticalExtents(vararg verticalExtent: VerticalExtent): GeospatialCoverageBuilder {
```
❓ Same question about List overload.

AzGoalie replied:
Must have been an oversight, good catch

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/LinkBuilder.kt:
```diff
@@ -0,0 +1,48 @@
+fun link(init: LinkBuilder.() -> Unit) = LinkBuilder().apply(init).build()
+
+class LinkBuilder : Builder<Link> {
+    private lateinit var href: String
```
❓ Are the errors clear enough to the consumer of the api when they try to invoke the `build` method on one of the builders where the `lateinit` vars have not been set? Are they obtuse or can the user figure out what they've done wrong?

AzGoalie replied:
The error returned is `kotlin.UninitializedPropertyAccessException: lateinit property href has not been initialized` which isn't vague, but also not the same as the other errors. I added a  check in `build`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/SecurityAttributeBuilder.kt:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    fun declassDate(declassDate: String): SecurityAttributeBuilder {
+        this.declassDate = declassDate
```
❓ Not a `Date`, either internally or as the parameter passed in? No need to overload to support that as input?

AzGoalie replied:
Changed to use new `DdmsDate` class.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/SecurityAttributeBuilder.kt:
```diff
@@ -0,0 +1,208 @@
+        return this
+    }
+
+    fun dateOfExemptedSource(dateOfExemptedSource: String): SecurityAttributeBuilder {
```
❓ Same question about operating on `Date` objects.

AzGoalie replied:
Changed to use new `DdmsDate` class.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/geospatial/PostalAddressBuilder.kt:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    fun streets(street: List<String>): PostalAddressBuilder {
+        streets.addAll(street)
```
❓ This hadn't occurred to me before now, but for all these builder methods that take varargs/lists, do you care how many times they are invoked by callers? Should you clear out the list on each invocation before adding the input contents?

AzGoalie replied:
I believe most builder objects like these don't clear the list. If we feel like it would make sense to, i can add it.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/geospatial/PostalAddressBuilder.kt:
```diff
@@ -0,0 +1,78 @@
+    private var city: String = ""
+    private var stateProvince: StateProvince? = null
+    private var postalCode: String = ""
+    private var countryCode: CountryCode = CountryCode()
```
❓ Should this be `lateinit` instead? What is a "default"/empty `CountryCode`?

AzGoalie replied:
The spec allows a 'blank' `CountryCode` object, but it could probably be more strict. The writer checks to see if the country code object is the 'default' and excludes it if so.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/producers/PersonBuilder.kt:
```diff
@@ -0,0 +1,74 @@
+fun person(init: PersonBuilder.() -> Unit) = PersonBuilder().apply(init).build()
+
+class PersonBuilder : ProducerBuilder {
+    private val names: ArrayList<String> = arrayListOf()
```
❓ How do the multiple `names` relate to the singular `surname`?

AzGoalie replied:
According to the spec, it seems to be any name that's not the surname. First, Middle, and any other if necessary.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/producers/ServiceBuilder.kt:
```diff
@@ -0,0 +1,56 @@
+fun service(init: ServiceBuilder.() -> Unit) = ServiceBuilder().apply(init).build()
+
+class ServiceBuilder : ProducerBuilder {
+    private val names: ArrayList<String> = arrayListOf()
```
❓ Any reason to deal with these three common fields this way instead of converting `ProducerBuilder` to an abstract class?

AzGoalie replied:
Moved, good catch.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,427 @@
+                            nextTag()
+                            while (localName != "geographicIdentifier") {
+                                if (eventType != XMLStreamConstants.END_ELEMENT) {
+                                    when (localName) {
```
✏️ These lambdas are nesting rather deeply. Can we flatten out a bit by extracting functions for the work, thereby keeping the overall logic at the same cognitive level within any given method?

It would also be nice to reduce the 109-line method down to something manageable and easy to read.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,427 @@
+    private fun parseSubjectCoverage() {
+        ddms20Builder.subjectCoverage {
+            nextTag() // Subject
+            nextTag()
```
✏️ This second call to get the next tag is getting one of the coverage types? Can it be commented to make it clearer?

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,427 @@
+                            }
+                        }
+                        "boundingBox" -> {
+                            nextTag() // WestBL
```
I guess this answers my earlier question about the canonical order of compass points in ddms. Very odd order.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,427 @@
+            direction(RelatedResources.Direction.valueOf(getDdmsAttribute("direction").capitalize()))
+            securityAttributes(getSecurityAttributes())
+            nextTag()
+            while (localName != "relatedResources") {
```
❓ The case on the collection of related resources is camel-cased like above and on each contained `RelatedResource` is init-caps?

AzGoalie replied:
one of the oddities of DDMS 2.0. A singular `ReleatedResource` starts with a capital, while plural `relatedResources` is camel case

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/resource/Title.kt:
```diff
@@ -0,0 +1,33 @@
+    init {
+        with(securityAttributes) {
+            require(classification != Classification.NO_CLASSIFICATION) {
+                "ddms:title and ddms:subtitle must contain a ism:classification"
```
✏️ S/b "...contain an ism:classification"

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/resource/Title.kt:
```diff
@@ -0,0 +1,33 @@
+            require(classification != Classification.NO_CLASSIFICATION) {
+                "ddms:title and ddms:subtitle must contain a ism:classification"
+            }
+            require(ownerProducer.isNotEmpty() && ownerProducer.any { it.isNotBlank() }) {
```
✏️ Can remove the first test.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/security/Security.kt:
```diff
@@ -0,0 +1,30 @@
+    init {
+        with(securityAttributes) {
+            require(classification != Classification.NO_CLASSIFICATION) {
+                "ddms:security must contain a ism:classification"
```
✏️ S/b "...contain an ism:classification"

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/security/Security.kt:
```diff
@@ -0,0 +1,30 @@
+            require(classification != Classification.NO_CLASSIFICATION) {
+                "ddms:security must contain a ism:classification"
+            }
+            require(ownerProducer.isNotEmpty() && ownerProducer.any { it.isNotBlank() }) {
```
✏️ Can remove the first test.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/security/ism/Classification.kt:
```diff
@@ -0,0 +1,41 @@
+
+    companion object {
+        fun getEnum(value: String): Classification {
+            return valueOf(value.replace("-", "_"))
```
❓ Do we not need to get the value of `No classification` back? Is that just reserved for missing values?

AzGoalie replied:
It's reserved for missing values.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/security/ism/SecurityAttributeStrings.kt:
```diff
@@ -0,0 +1,41 @@
+package org.codice.ddms.v2.security.ism
+
+class SecurityAttributeStrings(securityAttributes: SecurityAttributes) {
+    val classification = securityAttributes.classification
```
✏️ I think it would be easier to read if the fields that were directly copied over were in one group and those that were joined lists were then in a second group, perhaps with a blank line between them.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/summary/Description.kt:
```diff
@@ -0,0 +1,33 @@
+    init {
+        with(securityAttributes) {
+            require(classification != Classification.NO_CLASSIFICATION) {
+                "ddms:description must contain a ism:classification"
```
✏️ S/b "...contain an ism:classification"

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/summary/Description.kt:
```diff
@@ -0,0 +1,33 @@
+            require(classification != Classification.NO_CLASSIFICATION) {
+                "ddms:description must contain a ism:classification"
+            }
+            require(ownerProducer.isNotEmpty() && ownerProducer.any { it.isNotBlank() }) {
```
✏️ Can remove the first test.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/summary/GeospatialCoverage.kt:
```diff
@@ -0,0 +1,38 @@
+    val verticalExtents: List<VerticalExtent>
+) {
+    init {
+        require(geographicIdentifiers.isNotEmpty() ||
```
✏️ I believe I asked a similar question w/r/t this class' builder: shouldn't the test in fact be that there is a valid, non-null entry in one of those lists?

AzGoalie replied:
That should be taken care of by each of the individual types. Can't create an invalid/empty list of these objects.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    private val ddms: DdmsResource,
+    private val writer: XMLStreamWriter
+) : DdmsWriter, XMLStreamWriter by writer {
+    private val ddmsPrefix = "ddms"
```
✏️ Consts or vals in a companion object.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeDescription() {
+        ddms.description?.run {
```
✏️ It only matters in terms of a) consistency and b) return type, but I tend to prefer `apply` over `run`. It doesn't make a big difference, but SAML-CTK used `apply` over `run` if we want to shoot for Codice consistency.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeLanguages() {
+        for (language in ddms.languages) {
```
✏️ Alternately: `ddms.languages.forEach {}`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeDates() {
+        ddms.dates?.run {
```
✏️ Same comment re: `apply` vs `run`.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeRights() {
+        ddms.rights?.run {
```
✏️ Same comment re: `apply` vs `run`.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeSources() {
+        for (source in ddms.sources) {
```
✏️ Alternately: `ddms.sources.forEach {}`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeTypes() {
+        for (type in ddms.types) {
```
✏️ Alternately: `ddms.types.forEach {}`

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+        }
+    }
+
+    private fun writeOrganization(organization: Organization) {
```
✏️ Other than the wrapping `ddmsElement`, this method is identical to `writeService`. Could refactor to a single method that takes that element name and a `Producer`.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeFormat() {
+        ddms.format?.run {
```
✏️ Same comment re: apply vs run.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeTemporalCoverages() {
+        for (temporalCoverage in ddms.temporalCoverages) {
```
✏️ Alternately a foreach

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeVirtualCoverages() {
+        for (virtualCoverage in ddms.virtualCoverages) {
```
✏️ Alternately a foreach

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    private fun writeGeospatialCoverages() {
+        for (geospatialCoverage in ddms.geospatialCoverages) {
```
✏️ Alternately a foreach

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,550 @@
+                addSecurityAttributes(relatedResources.securityAttributes)
+                ddmsAttribute("relationship", relatedResources.relationship)
+                when (relatedResources.direction) {
+                    RelatedResources.Direction.Outbound -> ddmsAttribute("direction", "outbound")
```
❓ Would it be better to have a `toString` or `toOutput` method of some such on the enum itself that could print out its name in lowercase? Then you could just have a single `ddmsAttribute("direction", relatedResources.direction.toOutput())` call here.

AzGoalie replied:
i like the `toString` method.

coyotesqrl commented on src/test/kotlin/org/codice/ddms/gml/v3/validation/PositionTest.kt:
```diff
@@ -0,0 +1,30 @@
+
+class PositionTest {
+    @Test
+    fun `a Position create correct xml`() {
```
✏️ "creates"

coyotesqrl commented:
@bakejeyner, @ahoffer, @brendan-hofmann, @andrewkfiedler, @gordocanchola, and @brjeter

Tagging to remind you all that your reviews on this PR have been requested.

AzGoalie committed:
Addressed comments and refactored
Updated gradle to version 5.2
  Added dependency locking
Updated kotlin to version 1.3.21
Updated spotless to version 3.17.0
Updated mockito to version 2.24.0
Fixed typo in DateFormat
Updated Position toString to be more kotlin like
Made SrsAttributes more testable
Moved builder lambads into companion objects
Changed id in PolygonBuilder to be initialized to empty string
Added require check to SrsAttributes requiring srsDimention to be positive
Added validation to Identifier
Added extra validation to Title
Rearranged variables in Ddms20ResourceBuilder
Added new methods to GeospatialCoverageBuilder to take a list to go along with the vararg methods
Changed ordering of BoundingBox to be north south east west
Added a check to LinkBuilder for href
Added validation to Links to have a non blank href
Added new DdmsDate class for handling dates
Updated Dates to use new DdmsDate class
Updated packages inside org.codice.ddms.v2.builder to reflect the same structure as the non builder classes
Updated spotless license not needing to use root.dir anymore
Added Detekt
Made CountryCode more strict, no more default, requires qualifier and value
Refactored ProducerBuilder to have common functions
Fixed typos and redundant checks
Added README

AzGoalie commented:
I had to suppress some detekt issues, they are all marked by `TODO` comments. Would like some input on them :)

AzGoalie committed:
Added buildSrc for deps and libs, Added license to build files

AzGoalie committed:
Removed Java code style and refactored common styles

coyotesqrl commented on .gitattributes:
```diff
@@ -0,0 +1,22 @@
+*       text=auto
+
+# Explicitly declare text files we want to always be normalized and converted 
+# to native line endings on checkout.
```
❓ Should `*.kt` be added to this list?

coyotesqrl commented on codice.license.kt:
```diff
@@ -0,0 +1,13 @@
+/**
```
✏️ Might want to consider the shorter license header we used for the [SAML-CTK](https://github.com/codice/saml-conformance/blob/master/codice.license.kt)

coyotesqrl commented on src/main/kotlin/org/codice/ddms/DdmsDate.kt:
```diff
@@ -0,0 +1,112 @@
+        dateString = date
+
+        if (!isUnknown() && !isNotApplicable()) {
+            parsedDate = datetimeFormatter.parse(date)
```
✏️ In the case of a date and not an unknown or N/A date, this ctor is going to attempt to parse it twice: once in the validation method and then again here. I don't have a great solution off the top of my head for this.

AzGoalie replied:
Fixed

coyotesqrl commented on src/main/kotlin/org/codice/ddms/DdmsDate.kt:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    constructor(date: TemporalAccessor) {
+        require(isValid(date)) {
```
✏️ Same problem with running the parse logic twice.

AzGoalie replied:
Fixed

coyotesqrl commented on src/main/kotlin/org/codice/ddms/DdmsDate.kt:
```diff
@@ -0,0 +1,112 @@
+
+    override fun hashCode(): Int {
+        var result = dateString.hashCode()
+        result = 31 * result + (parsedDate?.hashCode() ?: 0)
```
❗️ Your `equals` and `hashCode` methods are not equivalent, as this method includes the `parsedDate` in the generation of the hash code.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/SrsAttributes.kt:
```diff
@@ -0,0 +1,58 @@
+
+    private fun srsNameToString(): String {
+        return if (srsName.isNotBlank()) {
+            "srsName=\"$srsName\""
```
✏️ Raw strings (triple-quoted) might read better than these escaped quote marks.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/gml/v3/SrsAttributes.kt:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    override fun toString(): String {
+        return srsNameToString() +
```
✏️ Not sure I wouldn't use a string template for this instead of concatenation.

AzGoalie replied:
I choose to use this instead because the methods contain a bunch of conditional logic.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/producers/PersonBuilder.kt:
```diff
@@ -0,0 +1,43 @@
+    private var affiliation: String = ""
+
+    fun surname(surname: String): PersonBuilder {
+        this.surname = surname
```
❓ Since this property is defined as `lateinit` and not pre-filled to an empty string, would it be reasonable to assume that empty string is not a valid value for it? In that case, shouldn't there be a validation check here?

This same question applies throughout the project where we have `lateinit` Strings.

AzGoalie replied:
The spec states that there must be a `surname` element, but it has no rules for if the value can be blank or not. I'm not sure which is better: Following the spec exactly or making this slightly more restrictive.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/security/SecurityAttributeBuilder.kt:
```diff
@@ -0,0 +1,212 @@
+import org.codice.ddms.v2.security.ism.Classification
+import org.codice.ddms.v2.security.ism.SecurityAttributes
+
+@Suppress("TooManyFunctions") // TODO: Not sure how we can make this 'smaller'
```
📓 I don't think you should bother trying to break it up. Any breakup here would be artificial. This is just a builder that has a lot of functions.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/builder/summary/GeospatialCoverageBuilder.kt:
```diff
@@ -0,0 +1,119 @@
+import org.codice.ddms.v2.summary.geospatial.PostalAddress
+import org.codice.ddms.v2.summary.geospatial.VerticalExtent
+
+@Suppress("TooManyFunctions") // TODO: Not sure how we can make this 'smaller'
```
📓 Same. Let's just suppress and leave comments that the builders are building complex data objects and any breakup is artificial.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,475 @@
+                "Organization" -> organization {
+                    parseProducer("Organization")
+                }
+                "Person" -> person {
```
I'm torn here. Part of me would prefer to see these builders extracted out to methods but I'm not sure that makes things more readable vs. less.

AzGoalie replied:
I have no preference, but i moved them to be their own functions.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,475 @@
+        }
+    }
+
+    private fun <T : ProducerBuilder<T>> ProducerBuilder<T>.parseProducer(type: String) {
```
✏️ Might be best to move extension functions to their own section of this file.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/v2/writer/Ddms20XmlWriter.kt:
```diff
@@ -0,0 +1,543 @@
+private const val GML_NAMESPACE = "http://www.opengis.net/gml"
+private const val GML_PREFIX = "gml"
+
+@Suppress("LargeClass", "TooManyFunctions") // TODO: Could break this up into logical parts to make this 'smaller'
```
✏️ One possibility here is to create classes to hold extension functions for each of the subtypes to be written, but those get a bit messy as you'd have to give them access to the common functions to do the actual writing, like `ddmsAttribute`, etc. I think it would make the structure harder to track and probably isn't worth the effort.

AzGoalie replied:
I'm just going to move the extension functions to be together like you suggested in your previous comment.

AzGoalie replied:
I was thinking of making multiple "readers" for each part of the ddms document. But that would only remove a few methods and make it harder to follow. For now i'm just going to move the extension functions to be together like your previous comment.

coyotesqrl commented on src/main/kotlin/org/codice/ddms/xml/util/XmlUtils.kt:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @param documentedTag The name of what the next tag should be.
+ */
+@Suppress("UnusedPrivateMember") // TODO: I think we should keep suppressing this one, or we could log it
```
✏️ I think just log this `documentedTag`.

AzGoalie replied:
Moved it to `Ddms20XmlReader` and logged the argument.

coyotesqrl reviewed (changes requested):
Oops. I accidentally approved when I meant to request changes for the hash code/equals incompatibility.

brjeter commented on .gitattributes:
```diff
@@ -0,0 +1,22 @@
+# Set default behaviour, in case users don't have core.autocrlf set.
```
✏️ I think the README should probably be updated with more information. It's pretty bare right now. Some usage examples may be nice.

brjeter replied:
Speaking of usage examples, it may be a good idea to have a test for this library in a Java class, just to make sure there are no interoperability snags.
I realize this is a decent amount of work, so it could be a future improvement. Just a thought.

AzGoalie committed:
Refactored and fixed double parse of DdmsDate

AzGoalie committed:
Added tests for unknown and not applicable DdmsDates.

AzGoalie committed:
Added dokka documentation generation.

AzGoalie committed:
Added KDoc comments

bakejeyner commented on src/main/kotlin/org/codice/ddms/v2/summary/TemporalCoverage.kt:
```diff
@@ -0,0 +1,21 @@
+    val name: String = "Unknown",
+    val start: DdmsDate = DdmsDate.unknown,
+    val end: DdmsDate = DdmsDate.unknown
+)
```
❓ (Disclaimer: I know nothing about ddms) Would it be valuable to add an init block that requires the start date be  before the end date?

AzGoalie replied:
~That's a great idea. :+1:~
Actually DDMS doesn't specify a restriction on the date, and since we're currently using `TemporalAcsessor` and `String` for holding the date, it would be pretty complicated to make this check unless we swapped to holding the individual parts of the date/time. I'd like to get the initial pr out before trying something like that.

bakejeyner commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,493 @@
+        val datum = getDdmsAttribute("datum")
+        return VerticalDistance(elementText.toDouble(),
+                if (unit.isNotBlank()) UnitOfMeasure.valueOf(unit) else null,
+                if (datum.isNotBlank()) Datum.valueOf(datum) else null)
```
✏️ NitPick: Could also write these as ternaries:
```
                unit.isNotBlank() ? UnitOfMeasure.valueOf(unit) : null,
                datum.isNotBlank() ? Datum.valueOf(datum) : null)
```

AzGoalie replied:
Sadly, Kotlin doesn't have a ternary operator, this is their suggested solution. :frowning_face:

bakejeyner commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,493 @@
+    private fun nextTag(documentedTag: String): Int {
+        logger.trace("Expecting '$documentedTag' as next text")
+        return nextTag()
+    }
```
❓ Should this function be verifying that the next tag is the tag we are expecting (or log when we get a tag we don't expect)?

AzGoalie replied:
I made this tag so it would be clearer than the comments i had stating what the 'next' tag was. There are a few areas that `documentedTag` is a bunch of options that the tag could be like `person / organization / service`. I could change the name to fit better but not sure what it would be.

bakejeyner commented on src/test/kotlin/org/codice/ddms/gml/v3/validation/LinearRingTest.kt:
```diff
@@ -0,0 +1,43 @@
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">0.0</gml:pos>" +
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">0.0</gml:pos>" +
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">0.0</gml:pos>" +
+                "</gml:LinearRing>"
```
✏️ Could use a raw string:
```
        val xml = """<gml:LinearRing xmlns:gml="http://www.opengis.net/gml">
                    |<gml:pos xmlns:gml="http://www.opengis.net/gml">0.0</gml:pos>
                    |<gml:pos xmlns:gml="http://www.opengis.net/gml">0.0</gml:pos>
                    |<gml:pos xmlns:gml="http://www.opengis.net/gml">0.0</gml:pos>
                    |<gml:pos xmlns:gml="http://www.opengis.net/gml">0.0</gml:pos>
                    |<gml:pos xmlns:gml="http://www.opengis.net/gml">0.0</gml:pos>
                    |</gml:LinearRing>""".trimMargin()
```

AzGoalie replied:
We can't, the expected string doesn't have any new lines and the raw string adds them.

bakejeyner commented on src/test/kotlin/org/codice/ddms/gml/v3/validation/PointTest.kt:
```diff
@@ -0,0 +1,43 @@
+    fun `a valid Point creates correct xml`() {
+        val xml = "<gml:Point gml:id=\"id\" srsName=\"wgs84\" xmlns:gml=\"http://www.opengis.net/gml\">" +
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">0.0 0.0</gml:pos>" +
+                "</gml:Point>"
```
✏️ Could use a raw string (example in first comment of this type)

AzGoalie replied:
We can't, the expected string doesn't have any new lines and the raw string adds them.

bakejeyner commented on src/test/kotlin/org/codice/ddms/gml/v3/validation/PolygonTest.kt:
```diff
@@ -0,0 +1,51 @@
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">1.0 1.0</gml:pos>" +
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">2.0 2.0</gml:pos>" +
+                "<gml:pos xmlns:gml=\"http://www.opengis.net/gml\">3.0 3.0</gml:pos>" +
+                "</gml:LinearRing></gml:exterior></gml:Polygon>"
```
✏️ Could use a raw string (example in first comment of this type)

AzGoalie replied:
We can't, the expected string doesn't have any new lines and the raw string adds them.

bakejeyner commented on src/test/kotlin/org/codice/ddms/gml/v3/validation/PositionTest.kt:
```diff
@@ -0,0 +1,22 @@
+class PositionTest {
+    @Test
+    fun `a Position creates correct xml`() {
+        val xml = "<gml:pos srsName=\"wgs84\" xmlns:gml=\"http://www.opengis.net/gml\">0.0 0.0</gml:pos>"
```
✏️ Could use a raw string (example in first comment of this type)

AzGoalie replied:
Done

bakejeyner commented on src/test/kotlin/org/codice/ddms/gml/v3/validation/SrsAttributesTest.kt:
```diff
@@ -0,0 +1,27 @@
+class SrsAttributesTest {
+    @Test
+    fun `SrsAttributes create valid xml attributes`() {
+        val xml = "srsName=\"srsName\" srsDimension=\"2\" axisLabels=\"axis labels\" uomLabels=\"uom labels\""
```
✏️ Could use a raw string (example in first comment of this type)

AzGoalie replied:
Done

AzGoalie committed:
Converted some tests to use raw strings for better readability

AzGoalie committed:
Added missing nextTag documentation strings

brjeter commented on detekt.yml:
```diff
@@ -0,0 +1,527 @@
+    active: false
+  WildcardImport:
+    active: true
+    excludeImports: 'org.codice.compliance.SAMLSpecRefMessage.*,org.apache.cxf.rs.security.saml.sso.SSOConstants.*,java.util.*,kotlinx.android.synthetic.*'
```
✏️ I'm guessing you don't need some of these 😄

brjeter commented on src/main/kotlin/org/codice/ddms/xml/util/XmlUtils.kt:
```diff
@@ -0,0 +1,82 @@
+        setPrefix(prefix, uri)
+    }
+
+    writeNamespace(prefix, uri)
```
❓ I'm confused about why these extension functions are necessary. Is it because you want to call `setPrefix` even if `prefix` is null or empty string?

AzGoalie replied:
When writing a namespace with `XMLStreamWriter` it excepts you to have already called `setPrefix`. This was just a helper extension method for calling `setPrefix` if it wasn't previously set.

brjeter commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,495 @@
+    }
+
+    override fun getAttributeValue(namespaceURI: String?, localName: String): String {
+        return reader.getAttributeValue(namespaceURI, localName) ?: ""
```
❓ Would it be overkill to trace log the attribute's `localName` here?

brjeter commented on src/main/kotlin/org/codice/ddms/v2/reader/Ddms20XmlReader.kt:
```diff
@@ -0,0 +1,495 @@
+
+    private fun <T : ProducerBuilder<T>> ProducerBuilder<T>.parseProducer(type: String) {
+        nextTag("$type name/phone/email")
+        while (localName != type) {
```
❓ This strikes me as a dangerous way to loop over this. Is there any likelihood that the DDMS could be malformed enough to cause this loop to execute more times than it should?

AzGoalie replied:
If the XML is so malformed, it would read until the end of the file and then throw an exception. Not sure how we could change it.

brjeter commented on src/main/kotlin/org/codice/ddms/DdmsResource.kt:
```diff
@@ -0,0 +1,51 @@
+    val description: Description?
+    val languages: List<Language>
+    val dates: Dates?
+    val rights: Rights?
```
❓ Why are some of these fields nullable and some (e.g. `subjectCoverage`) aren't?

AzGoalie replied:
The nullable fields are fields that are optional in the DDMS spec.

brjeter commented on src/main/kotlin/org/codice/ddms/v2/builder/producers/OrganizationBuilder.kt:
```diff
@@ -0,0 +1,16 @@
+*/
+package org.codice.ddms.v2.builder.producers
+
+import org.codice.ddms.v2.resource.producers.Organization
```
✏️ Nit: It bothers my obsessive tendencies that there is parity between the builder package and corresponding "buildee" package for all except for this one, since the `producers` package is nested under `resource` in one but not the other.

AzGoalie replied:
Good catch, moved.

AzGoalie committed:
Moved producer builders
